cvmfs/export_plugin/fs_traversal_libcvmfs.cc

```diff
@@ -0,0 +1,238 @@
+  int res = cvmfs_close(handle->ctx, handle->fd);
+  if (res != 0) return -1;
+  handle->fd = -1;
+  return 0;
+}
+
+int libcvmfs_do_fread(
+  void *file_ctx,
+  char *buff,
+  size_t len,
+  size_t *read_len
+) {
+  struct libcvmfs_file_handle *handle =
+    reinterpret_cast<libcvmfs_file_handle *>(file_ctx);
+  ssize_t read = cvmfs_pread(handle->ctx, handle->fd, buff, len, handle->off);
+  if (read == -1)  {
+    return -1;
+  }
+  *read_len = read;
+  handle->off += *read_len;
+  return 0;
+}
+
+int libcvmfs_do_fwrite(void *file_ctx, const char *buff, size_t len) {
+  return -1;
+}
+
+void libcvmfs_do_ffree(void *file_ctx) {
+  struct libcvmfs_file_handle *handle =
+    reinterpret_cast<libcvmfs_file_handle *>(file_ctx);
+  if (handle->fd > 0) {
+    libcvmfs_do_fclose(file_ctx);
+  }
+  free(handle->path);
+  free(handle);
+}
+
+
+struct fs_traversal_context *libcvmfs_initialize(
+  const char *repo,
+  const char *base,
+  const char *data,
+  int num_threads,
+  const char *config) {
+  if (!repo) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "Repository name must be specified");
+    return NULL;
+  }
+  int retval = 0;
+
+  struct fs_traversal_context *result = new struct fs_traversal_context;
+  result->version = 1;
+  char* major = reinterpret_cast<char *>(malloc(MAX_INTEGER_DIGITS));
+  snprintf(major, MAX_INTEGER_DIGITS, "%d", LIBCVMFS_VERSION_MAJOR);
+  char* minor = reinterpret_cast<char *>(malloc(MAX_INTEGER_DIGITS));
+  snprintf(minor, MAX_INTEGER_DIGITS, "%d", LIBCVMFS_VERSION_MINOR);
+  char* rev = reinterpret_cast<char *>(malloc(MAX_INTEGER_DIGITS));
+  snprintf(rev, MAX_INTEGER_DIGITS, "%d", LIBCVMFS_REVISION);
+  size_t len = 3 + strlen(major) + strlen(minor) + strlen(rev);
+  char* lib_version = reinterpret_cast<char *>(malloc(len));
+  snprintf(lib_version, len, "%s.%s:%s", major, minor, rev);
+
+  result->lib_version = strdup(lib_version);
+
+  free(major);
+  free(minor);
+  free(rev);
+  free(lib_version);
+
+  result->size = sizeof(*result);
+  result->repo = strdup(repo);
+  result->base = NULL;
+  result->data = NULL;
+  result->config = NULL;
+
+  cvmfs_option_map *options_mgr = cvmfs_options_init();
+  if (config) {
+    result->config = strdup(config);
+  } else {
+    size_t len = 8 + strlen(result->repo);
+    char *def_config = reinterpret_cast<char *>(malloc(len*sizeof(char)));
+    snprintf(def_config, len, "%s.config",  result->repo);
+    result->config = strdup(def_config);
+    free(def_config);
+  }
+  retval = cvmfs_options_parse(options_mgr, result->config);
+  if (retval) {
+    LogCvmfs(kLogCvmfs, kLogStderr,
+      "CVMFS Options failed to parse from : %s", result->config);
+    return NULL;
+  }
+
+  // Override repository name even if specified
+  cvmfs_options_set(options_mgr, "CVMFS_REPOSITORY_NAME", repo);
+
+  // Override Mount dir if not specified in configuration
+  if (base && !cvmfs_options_get(options_mgr, "CVMFS_MOUNT_DIR")) {
+    cvmfs_options_set(options_mgr, "CVMFS_MOUNT_DIR", base);
```

Comment:
For these additional options, do we always want them to overwrite the config, or leave the original value?